### PR TITLE
Reproduce pre-industrial output

### LINKED ISSUE #120: Reproduce pre-industrial output
As per https://github.com/penguian/access-esm-build-gadi/pull/13 create optimized executables that when tested with https://github.com/ACCESS-NRI/access-esm1.5-configs/tree/pre-industrial -- with appropriate changes in executables -- produces bit-identical output when compared to the same experiment run with the executables from `/g/data/access/payu/access-esm/bin/coe/`.

### PR DESCRIPTION
**Pull request by penguian**
Closes #120 
The executables `cice4`, `mom5` and `um7` executables in the following configuration reproduce the output of https://github.com/ACCESS-NRI/access-esm1.5-configs/tree/pre-industrial https://github.com/ACCESS-NRI/access-esm1.5-configs/blob/0f769ae4338005f8ed3c6ce6478e004842d4a598/config.yaml
```
 spack find
-- linux-rocky8-x86_64_v4 / gcc@8.5.0 ---------------------------
bzip2@1.0.8  c-blosc@1.21.5  cmake@3.24.2  diffutils@3.9  gmake@4.3  hdf5@1.10.11  libaec@1.0.6  libiconv@1.17  lz4@1.9.4  netcdf-c@4.7.1  openmpi@4.0.2  pkgconf@1.9.5  snappy@1.1.10  zlib-ng@2.1.4  zstd@1.5.5

-- linux-rocky8-x86_64_v4 / intel@19.0.3.199 --------------------
cice4@access-esm1.5  dummygrib@1.0  fcm@2021.05.0  gcom4@access-esm1.5  gmake@4.4.1  mom5@access-esm1.5  netcdf-fortran@4.5.2  oasis3-mct@access-esm1.5  um7@access-esm1.5
==> 24 installed packages
```
except for 3 to 6 point differences in one field in each of 12 `output000/atmosphere/aiihca.paa1*` UM output files, as per the following example:
```
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
* CUMF-II Comparison Report *
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

File 1: access-esm1.5-configs.53/output000/atmosphere/aiihca.paa1dec
File 2: access-esm1.5-configs.70/output000/atmosphere/aiihca.paa1dec
Files DO NOT compare
  * 0 differences in fixed_length_header (with 7 ignored indices)
  * 1 field differences, of which 1 are in data

Compared 2113/2113 fields, with 2112 matches

Maximum RMS diff as % of data in file 1: 0.58832395008410732 (field 443)
Maximum RMS diff as % of data in file 2: 0.58662723084335489 (field 443)

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
* Field 443/2113 - CO2 LAND SURFACE FLUX     KG/M**2/S *
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
Lookup compares, data DOES NOT compare
Compared 64/64 lookup values.
File_1 lookup info:
  t1(0101/12/01 00:00:335)  lblev(0)/blev(0.0)  lbproc(128)
Data differences:
  Number of point differences  : 6/27840
  Maximum absolute difference  : 547025.73463143967
  RMS difference               : 3288.6607554994839
  RMS diff as % of file_1 data : 0.58832395008410732
  RMS diff as % of file_2 data : 0.58662723084335489
```
Is this difference significant?

### CODE REVIEW

**Comment by penguian** [comment]
@blimlim  I conducted my testing from the `/g/data/tm70/pcl851/src/ACCESS-NRI/access-esm1.5-configs` directory. If you look at `config.yaml` in that directory you will see that the executables used are from three different sets of directories:

1. The original executables:
```
$ grep "  exe:" config.yaml
      exe: /g/data/access/payu/access-esm/bin/coe/um7.3x
      exe: /g/data/access/payu/access-esm/bin/coe/mom5xx
      exe: /g/data/access/payu/access-esm/bin/coe/cicexx
   exe: /g/data/access/payu/access-esm/bin/mppnccombine
```
2. The executables created via https://github.com/penguian/access-esm-build-gadi :
```
$ grep access-esm-build-gadi config.yaml
      # exe: /g/data/tm70/pcl851/src/penguian/access-esm-build-gadi/bin/um_hg3.exe-20240620
      # exe: /g/data/tm70/pcl851/src/penguian/access-esm-build-gadi/bin/momxx
      # exe: /g/data/tm70/pcl851/src/penguian/access-esm-build-gadi/bin/cice4.1_access-mct-12p-20240620
   # exe: /g/data/tm70/pcl851/src/penguian/access-esm-build-gadi/bin/mppnccombine
```
3. The executables created from the `120-reproduce-pre-industrial-output` branch:
```
$ grep linux-rocky8 config.yaml
      # exe: /g/data/tm70/pcl851/src/ACCESS-NRI/release/linux-rocky8-x86_64_v4/intel-19.0.3.199/um7-access-esm1.5-tysx4miwgxa6nzyx7v5sxnbosfok4z7o/bin/um_hg3.exe
      # exe: /g/data/tm70/pcl851/src/ACCESS-NRI/release/linux-rocky8-x86_64_v4/intel-19.0.3.199/mom5-access-esm1.5-ogwxk6dpwoah5dkrsppuxuevqgtjt3fe/bin/fms_ACCESS-CM.x
      # exe: /g/data/tm70/pcl851/src/ACCESS-NRI/release/linux-rocky8-x86_64_v4/intel-19.0.3.199/cice4-access-esm1.5-uwzdgjdt7hc3wgvceuby52sp32tuxb4k/bin/cice_access_360x300_12x1_12p.exe
   # exe: /g/data/tm70/pcl851/src/ACCESS-NRI/release/linux-rocky8-x86_64_v4/intel-19.0.3.199/mom5-access-esm1.5-ogwxk6dpwoah5dkrsppuxuevqgtjt3fe/bin/mppnccombine.spack
```

**Comment by penguian** [comment]
Re-running with `/g/data/tm70/pcl851/src/ACCESS-NRI/access-esm1.5-configs/config.yaml` set to use the executables from `/g/data/access/payu/access-esm/bin` produces essentially the same differences in output. See `/scratch/tm70/pcl851/access-esm/archive/output000.53-70.diff` versus `/scratch/tm70/pcl851/access-esm/archive/output000.53-71.diff`.

**Comment by penguian** [comment]
@harshula The Ruff errors in [`packages/oasis3-mct/package.py`](https://github.com/ACCESS-NRI/spack-packages/blob/120-reproduce-pre-industrial-output/packages/oasis3-mct/package.py) at https://github.com/ACCESS-NRI/spack-packages/actions/runs/9638336022 etc. are intentional, and result from leaving the strings as `f` strings.

File: packages/oasis3-mct/package.py [hunk 1]
```diff
@@ -215,7 +222,10 @@ def edit(self, spec, prefix):
         config = {}
 
         # TODO: https://github.com/ACCESS-NRI/ACCESS-OM/issues/12
-        NCI_OPTIM_FLAGS = "-g3 -O2 -axCORE-AVX2 -debug all -check none -traceback"
+        if "@access-esm1.5" in self.spec:
+            NCI_OPTIM_FLAGS = "-g3 -O2 -march=cascadelake -axCORE-AVX512 -debug all -check none -traceback"
+        else:
+            NCI_OPTIM_FLAGS = "-g3 -O2 -axCORE-AVX2 -debug all -check none -traceback"
```
Comment (harshula) [inline packages/oasis3-mct/package.py:228]: Set the default `NCI_OPTIM_FLAGS`, as before. Then for each special case, override the variable.

**Comment by harshula** [comment]
Hi @penguian , If you allow a netcdf-c range of 4.7.1 to 4.7.4. Don't you need to allow a netcdf-fortran range of 4.5.1 to 4.5.2?
```
$ cat /apps/netcdf/4.7.1/lib/Intel/pkgconfig/netcdf-fortran.pc 
prefix=/apps/netcdf/4.7.1
exec_prefix=${prefix}
libdir=/apps/netcdf/4.7.1/lib/Intel
includedir=/apps/netcdf/4.7.1/include/Intel
ccompiler=gcc
fcompiler=ifort

Name: netcdf-fortran
Description: NetCDF Client Library for Fortran
URL: http://www.unidata.ucar.edu/netcdf
Version: 4.5.1
Requires.private: netcdf > 4.1.1
Libs: -L${libdir} -lnetcdff
Libs.private: -L${libdir} -lnetcdff -lnetcdf
Cflags: -I${includedir}

$ cat /apps/netcdf/4.7.4/lib/Intel/pkgconfig/netcdf-fortran.pc 
prefix=/apps/netcdf/4.7.4
exec_prefix=${prefix}
libdir=/apps/netcdf/4.7.4/lib/Intel
includedir=/apps/netcdf/4.7.4/include/Intel
ccompiler=gcc
fcompiler=ifort

Name: netcdf-fortran
Description: NetCDF Client Library for Fortran
URL: http://www.unidata.ucar.edu/netcdf
Version: 4.5.2
Requires.private: netcdf > 4.1.1
Libs: -L${libdir} -lnetcdff
Libs.private: -L${libdir} -lnetcdff -lnetcdf
Cflags: -I${includedir}
```

File: packages/mom5/package.py [hunk 2]
```diff
@@ -35,15 +35,23 @@ class Mom5(MakefilePackage):
...
     depends_on("mpi")
     depends_on("netcdf-fortran@4.5.2:")
-    depends_on("netcdf-c@4.7.4:")
+    depends_on("netcdf-c@4.7.1:")
 
     with when("@:access-esm0,access-esm2:"):
+        depends_on("netcdf-c@4.7.4:")
+        depends_on("netcdf-fortran@4.5.2:")
+        # Depend on virtual package "mpi".
+        depends_on("mpi")
```
Comment (harshula) [inline packages/mom5/package.py:44]: Needs a bit of clean-up.

File: packages/cice4/package.py [hunk 3]
```diff
@@ -55,8 +55,11 @@ def edit(self, spec, prefix):
 
         config = {}
 
-        istr = join_path((spec["oasis3-mct"].headers).cpp_flags, "psmile.MPI1")
-        ideps = ["oasis3-mct", "netcdf-fortran"]
+        istr = " ".join([
+                join_path((spec["oasis3-mct"].headers).cpp_flags, "psmile.MPI1"),
+                join_path((spec["oasis3-mct"].headers).cpp_flags, "pio"),
```
Comment (harshula) [inline packages/cice4/package.py:60]: pio?

File: packages/cice4/package.py [hunk 4]
```diff
@@ -90,9 +93,9 @@ def edit(self, spec, prefix):
         # Based on https://github.com/coecms/access-esm-build-gadi/blob/master/patch/Macros.Linux.raijin.nci.org.au-mct
         config["intel"] = """
 ifeq ($(DEBUG), yes)
-    FFLAGS     := -r8 -i4 -O0 -g -align all -w -ftz -convert big_endian -assume byterecl -no-vec -xHost -fp-model precise
+    FFLAGS     := -r8 -i4 -O0 -g -align all -w -ftz -convert big_endian -assume byterecl -no-vec -xCORE-AVX2 -fp-model precise
 else
-    FFLAGS     := -r8 -i4 -O2 -align all -w -ftz -convert big_endian -assume byterecl -no-vec -xHost -fp-model precise
+    FFLAGS     := -r8 -i4 -O2 -align all -w -ftz -convert big_endian -assume byterecl -no-vec -march=cascadelake -xCORE-AVX512 -fp-model precise
```
Comment (harshula) [inline packages/cice4/package.py:98]: Is the `-march=cascadelake` actually required?

File: packages/mom5/package.py [hunk 5]
```diff
@@ -59,14 +67,18 @@ def edit(self, spec, prefix):
         makeinc_path = join_path(srcdir, "bin", "mkmf.template.spack")
         config = {}
 
-        istr = join_path((spec["oasis3-mct"].headers).cpp_flags, "psmile.MPI1")
         # NOTE: The order of the libraries matters during the linking step!
         if "@access-esm1.5" in self.spec:
-            ideps = ["oasis3-mct"]
+            istr = " ".join([
+                    join_path((spec["oasis3-mct"].headers).cpp_flags, "psmile.MPI1"),
+                    join_path((spec["oasis3-mct"].headers).cpp_flags, "pio"),
```
Comment (harshula) [inline packages/mom5/package.py:74]: pio?

File: packages/mom5/package.py [hunk 6]
```diff
@@ -59,14 +67,18 @@ def edit(self, spec, prefix):
...
+                    join_path((spec["oasis3-mct"].headers).cpp_flags, "mct")])
+            ideps = ["netcdf-fortran"]
             ldeps = ["oasis3-mct", "netcdf-c", "netcdf-fortran"]
-            FFLAGS_OPT = "-O3 -debug minimal -xCORE-AVX2"
+            FFLAGS_OPT = "-O3 -debug minimal -march=cascadelake -xCORE-AVX512 -align array64byte"
```
Comment (harshula) [inline packages/mom5/package.py:78]: Is the `-march=cascadelake` actually required?

File: packages/um7/package.py [hunk 7]
```diff
@@ -31,8 +31,9 @@ class Um7(Package):
     depends_on("dummygrib", type=("build", "link"))
     depends_on("fcm", type="build")
     depends_on("gcom4@access-esm1.5+mpi", type=("build", "link"))
-    depends_on("mpi", type=("build", "run"))
-    depends_on("netcdf-fortran", type=("build", "link"))
+    depends_on("openmpi@4.0.2:4.1.0", type=("build", "run"))
+    depends_on("netcdf-fortran@4.5.2", type=("build", "link"))
```
Comment (harshula) [inline packages/um7/package.py:35]: `4.5.1:4.5.2`?

**Comment by harshula** [comment]
Hi @penguian , Is `-march=cascadelake` overridden by the Spack compiler wrapper when `granularity` is set to `generic`?

File: packages/oasis3-mct/package.py [hunk 8]
```diff
@@ -215,7 +222,10 @@ def edit(self, spec, prefix):
         config = {}
 
         # TODO: https://github.com/ACCESS-NRI/ACCESS-OM/issues/12
-        NCI_OPTIM_FLAGS = "-g3 -O2 -axCORE-AVX2 -debug all -check none -traceback"
+        if "@access-esm1.5" in self.spec:
+            NCI_OPTIM_FLAGS = "-g3 -O2 -march=cascadelake -axCORE-AVX512 -debug all -check none -traceback"
```
Comment (harshula) [inline packages/oasis3-mct/package.py:226]: Is the `-march=cascadelake` actually required?

File: packages/oasis3-mct/package.py [hunk 9]
```diff
@@ -15,17 +15,24 @@ class Oasis3Mct(MakefilePackage):
...
-    # Depend on virtual package "mpi".
-    depends_on("mpi")
+    with when("@:access-esm0,access-esm2:"):
+        depends_on("netcdf-c@4.7.4:")
+        depends_on("netcdf-fortran@4.5.2:")
+        # Depend on virtual package "mpi".
+        depends_on("mpi")
+    with when("@access-esm1.5"):
+        depends_on("netcdf-c@4.7.1:4.7.4%gcc")
+        depends_on("netcdf-fortran@4.5.2")
```
Comment (harshula) [inline packages/oasis3-mct/package.py:33]: `4.5.1:4.5.2`?

**Review by blimlim — DISMISSED** [review]
Thanks for working this all out. I don't have enough knowledge to comment on any of the  spack/compiler parts, but based on Martin's comments that the non-reproducible variables aren't relevant in the concentration driven runs, it sounds like everything is working properly to me!

File: packages/oasis3-mct/package.py (hunk 1 above)
Comment (penguian) [reply packages/oasis3-mct/package.py]: OK in general. In this instance there is only one special case, so leaving as-is for now.

File: packages/cice4/package.py [hunk 10]
```diff
@@ -20,10 +20,10 @@ class Cice4(MakefilePackage):
 
     version("access-esm1.5", branch="access-esm1.5")
 
-    # Depend on virtual package "mpi".
-    depends_on("mpi")
-    depends_on("netcdf-fortran@4.5.2:")
-    depends_on("netcdf-c@4.7.1:")
+    depends_on("netcdf-c@4.7.1:4.7.4%gcc")
+    depends_on("netcdf-fortran@4.5.2")
```
[the remark this answers is not in the document]
Comment (penguian) [reply packages/cice4/package.py]: Done in all instances. Please check.

File: packages/mom5/package.py (hunk 2 above)
Comment (penguian) [reply packages/mom5/package.py]: Done.

File: packages/mom5/package.py [hunk 11]
```diff
@@ -35,15 +35,23 @@ class Mom5(MakefilePackage):
...
+        # Depend on virtual package "mpi".
+        depends_on("mpi")
         depends_on("datetime-fortran")
         depends_on("oasis3-mct+deterministic", when="+deterministic")
         depends_on("oasis3-mct~deterministic", when="~deterministic")
         depends_on("libaccessom2+deterministic", when="+deterministic")
         depends_on("libaccessom2~deterministic", when="~deterministic")
     with when("@access-esm1.5"):
+        depends_on("netcdf-c@4.7.1:4.7.4%gcc")
+        depends_on("netcdf-fortran@4.5.2")
```
[the remark this answers is not in the document]
Comment (penguian) [reply packages/mom5/package.py]: Done in all instances. Please check.

File: packages/cice4/package.py (hunk 3 above)
Comment (penguian) [reply packages/cice4/package.py]: Removed in all instances. Please check.

File: packages/cice4/package.py (hunk 4 above)
Comment (penguian) [reply packages/cice4/package.py]: Removed in all instances. Please check.

File: packages/mom5/package.py (hunk 5 above)
Comment (penguian) [reply packages/mom5/package.py]: Removed in all instances. Please check.

File: packages/mom5/package.py (hunk 6 above)
Comment (penguian) [reply packages/mom5/package.py]: Removed in all instances. Please check.

File: packages/um7/package.py (hunk 7 above)
Comment (penguian) [reply packages/um7/package.py]: Done in all instances. Please check.

File: packages/oasis3-mct/package.py (hunk 8 above)
Comment (penguian) [reply packages/oasis3-mct/package.py]: Removed in all instances. Please check.

File: packages/oasis3-mct/package.py (hunk 9 above)
Comment (penguian) [reply packages/oasis3-mct/package.py]: Done in all instances. Please check.

**Comment by penguian** [comment]
With packages as follows:
```
$ spack find
-- linux-rocky8-x86_64_v4 / gcc@8.5.0 ---------------------------
bzip2@1.0.8  cmake@3.24.2  diffutils@3.9  gmake@4.3  libaec@1.0.6  libiconv@1.17  lz4@1.9.4  pkgconf@1.9.5  snappy@1.1.10  zstd@1.5.5

-- linux-rocky8-x86_64_v4 / intel@19.0.3.199 --------------------
cice4@access-esm1.5  dummygrib@1.0  gcom4@access-esm1.5  hdf5@1.10.11        netcdf-c@4.7.4        oasis3-mct@access-esm1.5  pkgconf@1.9.5      zlib-ng@2.1.4
cmake@3.24.2         fcm@2021.05.0  gmake@4.4.1          mom5@access-esm1.5  netcdf-fortran@4.5.2  openmpi@4.1.0             um7@access-esm1.5
==> 25 installed packages
```
the output of the `pre-industrial` experiment is still essentially bitwise identical to the output of https://github.com/ACCESS-NRI/access-esm1.5-configs/tree/pre-industrial https://github.com/ACCESS-NRI/access-esm1.5-configs/blob/0f769ae4338005f8ed3c6ce6478e004842d4a598/config.yaml which used the `coe` executables.

**Comment by penguian** [comment]
The `um7` SPD still allows `-xHost`, and also has a strange structure with many explicit environment variable definitions. I will create a separate Issue to address this. Even so, I think that we are ready to try another iteration of https://github.com/ACCESS-NRI/ACCESS-ESM1.5/blob/2-spack-yaml/spack.yaml

File: packages/oasis3-mct/package.py (hunk 1 above)
Comment (harshula) [reply packages/oasis3-mct/package.py]: No, there are three:
```
        if "@access-esm1.5" in self.spec:
            NCI_OPTIM_FLAGS = "-g3 -O2 -xCORE-AVX512 -debug all -check none -traceback"
        else:
            NCI_OPTIM_FLAGS = "-g3 -O2 -axCORE-AVX2 -debug all -check none -traceback"
        CFLAGS = ""
        if "+deterministic" in self.spec:
            NCI_OPTIM_FLAGS = "-g0 -O0 -axCORE-AVX2 -debug none -check none"
            CFLAGS = "-g0"

        if "+optimisation_report" in self.spec:
            NCI_OPTIM_FLAGS += " -qopt-report=5 -qopt-report-annotate"
```

**Comment by harshula** [comment]
Hi @penguian , The `netcdf-fortran` SPD contains:
```
    depends_on("netcdf-c")
    depends_on("netcdf-c@4.7.4:", when="@4.5.3:")  # nc_def_var_szip required
    depends_on("doxygen", when="+doc", type="build")
```
If the source package does not contain any `C` code and only contains `Fortran` code, just depend on `netcdf-fortran`.

File: packages/oasis3-mct/package.py (hunk 1 above)
Comment (penguian) [reply packages/oasis3-mct/package.py]: Addressed. Please check.

**Comment by penguian** [comment]
I changed `packages/cice4/package.py` as `cice4` does not use C in any meaningful way.

**Review by harshula — DISMISSED** [review]

**Comment by penguian** [comment]
@harshula Thanks for the approval, but this PR was still a draft. I have now completed my intended changes.

File: packages/um7/package.py [hunk 12]
```diff
@@ -13,6 +13,21 @@
 import os
 import os.path
 
+def _bld_cfg_path(opt_value, hg):
```
Comment (harshula) [inline packages/um7/package.py:16]: Why is this a non-member function? You can make it a member function of the class `Um7` and then call it from within the class by doing `self._bld_cfg_path(...)`

File: packages/um7/package.py [hunk 13]
```diff
@@ -13,6 +13,21 @@
...
+    "opt" value (opt_value) and HadGEM number (hg).
+    """
+    if opt_value == "debug":
+        bld_config = f"bld-dbg-hadgem{hg}-C2.cfg"
+    else:
+        bld_config = f"bld-hadgem{hg}-mct.cfg"
+    return join_path(f"ummodel_hg{hg}", "cfg", bld_config)
+
+
+_hg = 3  # build HadGEM3 ONLY here
```
Comment (harshula) [inline packages/um7/package.py:28]: Make it a member variable of the class `Um7`.

File: packages/um7/package.py [hunk 14]
```diff
@@ -79,24 +112,18 @@ def install(self, spec, prefix):
...
         # Build with fcm
-        fcm("build", "-f", "-j", "4",
-            join_path(bld_dir, "cfg", bld_config))
+        fcm("build", "-f", "-j", "4", _bld_cfg_path(opt_value, _hg))
 
         # Install
         mkdirp(prefix.bin)
         install(
-            join_path(bld_dir, "bin", um_exe),
+            join_path(f"ummodel_hg{_hg}", "bin", um_exe),
```
Comment (harshula) [inline packages/um7/package.py:127]: Perhaps use a member function to create this string. `_bld_cfg_path` can then call the function too.

File: packages/um7/package.py (hunk 12 above)
Comment (penguian) [reply packages/um7/package.py:16]: I had already tried using a member function and didn't like the `self` references. I will change it back.

File: packages/um7/package.py (hunk 13 above)
Comment (penguian) [reply packages/um7/package.py:28]: I will just hardcode the 3 since it will never change.

**Comment by penguian** [comment]
Testing in progress. Converting to draft.

File: packages/um7/package.py (hunk 14 above)
Comment (penguian) [reply packages/um7/package.py]: Done.

File: packages/um7/package.py (hunk 13 above)
Comment (penguian) [reply packages/um7/package.py]: Done.

File: packages/um7/package.py (hunk 12 above)
Comment (penguian) [reply packages/um7/package.py]: Done.

**Review by harshula — APPROVED** [review]

**Pull request merged**